File: app/model/tools/ebnf/conversion/GrammarConverter.scala
package model.tools.ebnf.conversion

import model.tools.ebnf.grammar.{Grammar, GrammarElement, Variable}

final case class GrammarElemConvResult[IE <: GrammarElement, OE <: GrammarElement](
  newOutputElement: OE,
  newVariables: Seq[Variable] = Seq.empty,
  newReplacers: Map[IE, Variable] = Map.empty[IE, Variable],
  newRules: Seq[(Variable, IE)] = Seq.empty
)

final case class CollectionGrammarElementConvResult[IE <: GrammarElement, OE <: GrammarElement](
  newOutputElements: Seq[OE],
  newVariables: Seq[Variable] = Seq.empty,
  newReplacers: Map[IE, Variable] = Map.empty[IE, Variable],
  newRules: Seq[(Variable, IE)] = Seq.empty
)

trait GrammarConverter {

  protected type IE <: GrammarElement
  protected type IG <: Grammar[IE]
  protected type IR = (Variable, IE)

  protected type OE <: GrammarElement
  protected type OG <: Grammar[OE]
  protected type OR = (Variable, OE)

  def convertElement(
    element: IE,
    currentVariable: Seq[Variable] = Seq.empty,
    currentReplacers: Map[IE, Variable] = Map.empty,
    ruleVariable: Option[Variable] = None
  ): GrammarElemConvResult[IE, OE]

  def convert(grammar: IG): OG = {

    @annotation.tailrec
    def go(
      inputRules: List[IR],
      outputRules: Seq[OR],
      currentNonTerminals: Seq[Variable],
      currentReplacers: Map[IE, Variable]
    ): OG = inputRules match {
      case Nil => createOutputGrammar(grammar.startSymbol, outputRules.toMap)
      case (left, right) :: tail =>
        val GrammarElemConvResult(newRight, newNonTerminals, newReplacers, newRules) =
          convertElement(right, currentNonTerminals, currentReplacers, Some(left))

        go(
          tail ++ newRules,
          outputRules :+ (left -> newRight),
          currentNonTerminals ++ newNonTerminals,
          currentReplacers ++ newReplacers
        )
    }

    go(grammar.rules.toList, Seq.empty, grammar.variables, Map.empty)
  }

  protected def createOutputGrammar(startSymbol: Variable, rules: Map[Variable, OE]): OG

  def findNewVariable(currentTerminals: Seq[Variable]): Variable = {
    val countedSymbols = for {
      maybeIndex <- Seq(None) ++ (0 to 9).map(Some.apply)
      letter     <- 'A' to 'Z'
    } yield Variable(letter.toString, maybeIndex)

    countedSymbols.find(newNonTerm => !currentTerminals.contains(newNonTerm)).get
  }

}

File: react-ui/src/collectionTools/tools/ebnf/grammarConverter.ts
import {getVariablesFromGrammar, Grammar, GrammarRule} from './grammar';
import {ExtendedBackusNaurFormGrammarElement, variable, Variable} from './ebnfElements';

export type Replacers<IE> = [IE, Variable][];

interface GrammarElemConvResult<IE, OE> {
  newRules: GrammarRule<IE>[];
  updatedVariables: Variable[];
  updatedReplacers: Replacers<IE>;
}

export interface GrammarElementConversionResult<IE, OE> extends GrammarElemConvResult<IE, OE> {
  newOutputElement: OE;
}

export interface CollectionGrammarElementConversionResult<IE, OE> extends GrammarElemConvResult<IE, OE> {
  newOutputElements: OE[];
}


export abstract class GrammarConverter<IE extends ExtendedBackusNaurFormGrammarElement, IG extends Grammar<IE>, OE, OG extends Grammar<OE>> {

  protected abstract convertElement(
    element: IE,
    currentVariables: Variable[],
    currentReplacers: Replacers<IE>,
    ruleVariable: Variable | undefined
  ): GrammarElementConversionResult<IE, OE>;

  protected abstract createOutputGrammar(startSymbol: Variable, rules: GrammarRule<OE>[]): OG;

  private goConvertGrammar(
    inputRules: GrammarRule<IE>[],
    outputRules: GrammarRule<OE>[],
    currentVariables: Variable[],
    currentReplacers: Replacers<IE>
  ): GrammarRule<OE>[] {
    if (inputRules.length === 0) {
      return outputRules;
    } else {

      const [{variable, right}, ...tail] = inputRules;

      const {newOutputElement, newRules, updatedVariables, updatedReplacers} = this.convertElement(right, currentVariables, currentReplacers, variable);

      return this.goConvertGrammar(
        [...tail, ...newRules],
        [...outputRules, {variable: variable, right: newOutputElement}],
        [...currentVariables, ...updatedVariables],
        updatedReplacers
      );
    }
  }

  convert({rules, startSymbol}: IG): OG {
    const variables: Variable[] = getVariablesFromGrammar({rules, startSymbol});

    const newRules = this.goConvertGrammar(rules, [], variables, []);

    return this.createOutputGrammar(startSymbol, newRules);
  }

  protected findNewVariables(currentVariables: Variable[]): Variable | undefined {

    const vars = Array.from(new Set(currentVariables.map(({value}) => value)));

    const foundVar = Array.from({length: 26}, (_, i) => String.fromCharCode('A'.charCodeAt(0) + i))
      .find((s) => !vars.includes(s));

    return foundVar ? variable(foundVar) : undefined;
  }

}
